File: Waterrunoff/WaterRunoff.R
library(tidyverse)

water <- read.table("https://mheaton.byu.edu/Courses/Stat330/HomeworkAnalyses/3%20-%20WaterRunoff/Data/water.txt",sep = "", header = TRUE)

prec <- water$Precip

runf <- water$Runoff

# Check assumption and fit model
ggplot(water, aes(x=prec, y=runf)) + geom_point() + xlab("Precipitation") + ylab("Water Avilability")

cor(prec,runf)

cor(prec,runf)^2

slr <- lm(runf~prec, data = water)

library(lmtest)

bptest(slr)

library(MASS)

std <- stdres(slr)

hist(std)

ks.test(std,"pnorm")

plot(slr$residuals ~ slr$fitted.values)

abline(h=0)

ggplot(water, aes(x=prec, y=runf)) + geom_point() + geom_smooth(method = "lm", se=FALSE)

summary(slr)

confint(slr,level=.95)

# Add prediction line
ggplot(water,aes(x=prec, y=runf))+
  geom_point() + geom_smooth(method = "lm", se=FALSE)
pred_d <- data.frame(prec =seq(min(prec),max(prec),len = 1000))
pred <- predict.lm(slr, newdata = pred_d, interval = "prediction")
pred <- as.data.frame(pred)
ggplot(water, aes(x=prec,y=runf))+geom_point()+geom_smooth(method="lm",se=FALSE) + 
  geom_line(data=pred,aes(x=pred_d$prec,y=lwr),col="red") + 
  geom_line(data=pred,aes(x=pred_d$prec,y=upr),col="red")

#cross validation
n.cv <- 250
bias <- rep(NA,n.cv)
rpmse <- rep(NA,n.cv)
cvg <- rep(NA,n.cv)
wid <- rep(NA,n.cv)
n.test <- 4

for(i in 1:n.cv){
  #split into test and training set
  test.obs <- sample(1:nrow(water),n.test)
  test.set <- water[test.obs,]
  train.set <- water[-test.obs,]
  
  # fit a lm using training data only
  train.lm <- lm(Runoff ~ Precip, data=train.set)
  
  # Prediction and prediction intervals
  test.pred <- predict.lm(train.lm,newdata = test.set,interval="prediction",level = 0.95)  
  
  # calculate results
  bias[i] <- mean(test.pred[,1] - test.set$Runoff)
  rpmse[i] <- sqrt(mean((test.pred[,1] - test.set$Runoff)^2))
  cvg[i] <- mean(test.pred[,2] < test.set$Runoff & test.pred[,3] > test.set$Runoff)
  wid[i] <- mean(test.pred[,3]-test.pred[,2])  
}
mean(bias)
mean(rpmse)
mean(cvg)
mean(wid)

predict.lm (slr,newdata=data.frame(prec=c(4.5)), interval="prediction", level = 0.95)

File: Waterrunoff/waterRunoff.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.formula.api as smf
import statsmodels.stats.api as sm

# Reading in Data
water = pd.read_table("/Users/jungeolkim/Desktop/WD/venv/water.txt", sep = " ")
print(water)
water.shape

# Exploratory Data Analysis
sns.regplot(x = 'Precip', y = 'Runoff', data = water, color = 'blue', lowess = True)
sns.set()
plt.title('Water runoff by Precipitation')
plt.xlabel('Precipitation')
plt.ylabel('Water runoff')

water.corr()

# Fitting a linear model
mylm = smf.ols(formula = 'Runoff ~ Precip', data = water).fit()
mylm.summary()
mylm.mse_resid
mylm.params
mylm.rsquared

# Linearity
sns.regplot(x = 'Precip', y = 'Runoff', data = water, color = 'blue', ci = None)

# Plot of residuals vs. fitted values
dir(mylm)

sns.regplot(x = mylm.fittedvalues, y = mylm.resid, color = 'blue', ci = None)

sm.het_breuschpagan(mylm.resid_pearson, mylm.model.exog)

# Normality
sns.distplot(mylm.resid_pearson, kde = False)
mylm.summary()

# Hypothesis test
mylm.summary()

# Cross Validation
ncv = 250
bias = np.repeat(np.NaN,ncv)
rpmse = np.repeat(np.NaN,ncv)
wid = np.repeat(np.NaN,ncv)
ntest = 4
for cv in range(ncv) :
    # Choose which obs to put
    testobs = np.random.choice(water.shape[0],ntest)
    
    # Split data
    testset = water.iloc[testobs,:]
    trainset = water.drop(testobs)
    
    # Using training data to fit a model
    trainlm = smf.ols(formula = 'Runoff ~ Precip', data = trainset).fit()
    
    # Predict test set
    testpreds = trainlm.get_prediction(testset)
    
    # Calculate bias
    bias[cv] = np.mean(testpreds.predicted_mean-testset['Runoff'])
    
    # Calculate RPMSE
    rpmse[cv] = np.sqrt(np.mean(np.square(testpreds.predicted_mean-testset['Runoff'])))

    # Calculate Coverage
    pis = testpreds.conf_int(obs=True, alpha = 0.05)
    conditions_array = [pis[:,0] <= testset['Runoff'].values, pis[:,1] >= testset['Runoff'].values]
    cond_met = np.all(conditions_array,axis=0)
    
    # Calculate Width
    wid[cv] = np.mean(pis[:,1] - pis[:,0])

np.mean(bias)
np.mean(rpmse)
np.mean(cond_met)
np.mean(wid)

# 95% confidence intervals
mylm.conf_int(alpha=0.05)

# 95% prediction intervals for Precip at 4.5
predframe = pd.DataFrame(dict(Precip = [4.5]))

preds = mylm.get_prediction(predframe)

preds.summary_frame()
preds.predicted_mean
preds.conf_int(obs = True, alpha = 0.05)